storage: Avoid replica thrashing when localities are different sizes

a-robinson opened the pull request:
Fixes #20241

Release note (bug fix): avoid rebalance thrashing when localities have
very different numbers of nodes

This could use some careful review. A more comprehensive refactoring as described in #20751 would be nice, but this fixes the immediate problem as seen in #20241.

a-robinson committed:
storage: make diversity scoring tests make more compact

Release note: None

a-robinson committed:
storage: Use MaxDiversityScore constant where appropriate

Release note: None

a-robinson committed:
storage: Always simulate RemoveTarget when rebalancing

Skipping the simulation when raftStatus.Progress is nil can make
for undesirable thrashing of replicas, as seen when testing #20241.
It's better to run the simulation without properly filtering replicas
than to not run it at all.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20752)
<!-- Reviewable:end -->

a-robinson committed:
storage: Avoid replica thrashing when localities are different sizes

Fixes #20241

Release note (bug fix): avoid rebalance thrashing when localities have
very different numbers of nodes

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 540 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0UrowrWZ9hckDClkqF:-L0UrowrWZ9hckDClkqG:by37jct) ([raw file](https://github.com/cockroachdb/cockroach/blob/afa184d23c4474508f5c0f19870f3fbf573ca04c/pkg/storage/allocator.go#L540)):*
> ```Go
> 
> 		replicaCandidates := desc.Replicas
> 		if raftStatus != nil && raftStatus.Progress != nil {
> ```

This could use a comment. When are `raftStatus` and `raftStatus.Progress` nil?

---

*[pkg/storage/allocator.go, line 582 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0UsfJauAxMYCMTEV_h:-L0UsfJauAxMYCMTEV_i:b-3r8wf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff35505bbcfc245da2015761496d4c3c1d40a93b/pkg/storage/allocator.go#L582)):*
> ```Go
> }
> 
> func shouldRebalanceToFrom(
> ```

This name isn't terribly descriptive, but I can't think of a better name. A comment describing what it does would be helpful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-:-L0UtTXcRDG_XjjHdIki:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 582 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0UsfJauAxMYCMTEV_h:-L0WmL-wOmuJhcvd2Yx4:bgtxchb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff35505bbcfc245da2015761496d4c3c1d40a93b/pkg/storage/allocator.go#L582)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This name isn't terribly descriptive, but I can't think of a better name. A comment describing what it does would be helpful.
</blockquote></details>

`Pair` might be clearer than `ToFrom`, which raised a parse error in my brain. 

---

*[pkg/storage/allocator.go, line 591 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0WmYxkiGUoFfoSG8ya:-L0WmYxkiGUoFfoSG8yb:b-71zos0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff35505bbcfc245da2015761496d4c3c1d40a93b/pkg/storage/allocator.go#L591)):*
> ```Go
> ) bool {
> 	if from.StoreID == to.store.StoreID {
> 		log.VEventf(ctx, 2, "not rebalancing to s%d because we'd immediately remove it: %s",
> ```

Shouldn't equal store IDs be caught earlier in the process? This is the "can't have two replicas per store (or even per node)" rule, not "we'd immediately remove it". 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-:-L0WnmpgmYJ2gVHRGDmc:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a6802739 reviewed (commented):
LGTM!

a6802739 commented on pkg/storage/allocator_scorer.go:
```diff
@@ -136,6 +136,38 @@ func (c candidate) less(o candidate) bool {
+	// balanceScore ties and it's a workable stop-gap on the way to something
+	// like #20751.
+	avgRangeCount := float64(c.rangeCount+o.rangeCount) / 2.0
+	overfullThreshold := math.Max(overfullRangeThreshold(options, avgRangeCount), avgRangeCount+1.5)
```
Could you please explain the definition of `overfullThreshold` here? Thank you very much.

a6802739 reviewed (commented):
LGTM

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-:-L0fSV6qCfcUAP97xlaO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
TFTRs!

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/allocator.go, line 540 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0UrowrWZ9hckDClkqF:-L0fYIfEQhl2Itfld-Fg:bmyyaq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/afa184d23c4474508f5c0f19870f3fbf573ca04c/pkg/storage/allocator.go#L540)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This could use a comment. When are `raftStatus` and `raftStatus.Progress` nil?
</blockquote></details>

I saw `raftStatus.Progress` being nil repeatedly when reproducing #20241. I never saw `raftStatus` being nil. I didn't dig deep into it at the time, but I'd assume it's because there was a leader-not-leaseholder situation. Do you think it's worth more investigation?

Added a comment, either way.

---

*[pkg/storage/allocator.go, line 582 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0UsfJauAxMYCMTEV_h:-L0fZUJP0DjkDyBVjXC2:b-5mb18r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff35505bbcfc245da2015761496d4c3c1d40a93b/pkg/storage/allocator.go#L582)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`Pair` might be clearer than `ToFrom`, which raised a parse error in my brain. 
</blockquote></details>

Changed to `shouldRebalanceBetween` and added a comment.

---

*[pkg/storage/allocator.go, line 591 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0WmYxkiGUoFfoSG8ya:-L0fZoHYg_3equ6E7-Kq:b-1qzpxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff35505bbcfc245da2015761496d4c3c1d40a93b/pkg/storage/allocator.go#L591)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Shouldn't equal store IDs be caught earlier in the process? This is the "can't have two replicas per store (or even per node)" rule, not "we'd immediately remove it". 
</blockquote></details>

My bad variable naming confused you. The variable previously named `from` isn't necessarily an existing member of the range -- it's the replica that we think we'll remove after adding the new replica. This is here to prevent us from adding a replica that we would then just immediately remove.

---

*[pkg/storage/allocator_scorer.go, line 167 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0QJ-NhdMR3dU04Iohu-r4-167:-L0f_fJRiMkcDiw_8Ppu:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff35505bbcfc245da2015761496d4c3c1d40a93b/pkg/storage/allocator_scorer.go#L167)):*
<details><summary><i>Previously, a6802739 (songhao) wrote…</i></summary><blockquote>

Could you please explain the definition of `overfullThreshold` here? Thank you very much.
</blockquote></details>

Added a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-:-L0faDeDcAMJC7RcOawK:bz5mair)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 540 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752#-L0UrowrWZ9hckDClkqF:-L0ffp0WRwv2ZVFMa3NB:b-nq27un) ([raw file](https://github.com/cockroachdb/cockroach/blob/afa184d23c4474508f5c0f19870f3fbf573ca04c/pkg/storage/allocator.go#L540)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I saw `raftStatus.Progress` being nil repeatedly when reproducing #20241. I never saw `raftStatus` being nil. I didn't dig deep into it at the time, but I'd assume it's because there was a leader-not-leaseholder situation. Do you think it's worth more investigation?

Added a comment, either way.
</blockquote></details>

Yes, that sounds like leader-not-leaseholder. Progress is only present on the leader.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20752)*
<!-- Sent from Reviewable.io -->
